### PR TITLE
Refactor Sapling components and builder into `zcash_primitives::sapling`

### LINKED ISSUE #736: Internally refactor Sapling tx components and builder to `zcash_primitives::sapling`
As part of the effort to recreate the `sapling-crypto` crate, we want to collect all of its code under `zcash_primitives::sapling`. Some of it currently lives under `zcash_primitives::transaction::components::sapling`, and we should refactor it to match the `orchard::{bundle, builder, ...}` layout. We can do this by moving them into `pub(crate)` submodules, and re-exporting the types in their current locations (so the `zcash_primitives` API is unaffected).

### PR DESCRIPTION
**Pull request by str4d**
Closes #736.

### CODE REVIEW

**Comment by str4d** [comment]
Best reviewed locally with `git show COMMIT --color-moved=zebra`.

**Comment by str4d** [comment]
I haven't moved `zcash_primitives::transaction::components::sapling::fees` yet, because I'm not entirely sure how to handle it (as the traits directly use amount types that will not be moving into `sapling-crypto`).

**Comment by nuttycom** [comment]
> I haven't moved `zcash_primitives::transaction::components::sapling::fees` yet, because I'm not entirely sure how to handle it (as the traits directly use amount types that will not be moving into `sapling-crypto`). I suspect it will stay where it is?

Yeah, I think keeping this where it is makes sense for now. TBH these traits could probably be renamed and just moved (along with the sibling traits for transparent and TZE) into `zcash_primitives::transaction::fees` itself.

File: zcash_primitives/src/transaction/components/sapling.rs [hunk 1]
```diff
@@ -477,196 +103,39 @@ impl SpendDescription<Authorized> {
...
-impl<A> std::fmt::Debug for OutputDescription<A> {
-    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> Result<(), std::fmt::Error> {
-        write!(
-            f,
-            "OutputDescription(cv = {:?}, cmu = {:?}, ephemeral_key = {:?})",
-            self.cv, self.cmu, self.ephemeral_key
-        )
+        Ok(SpendDescriptionV5::from_parts(cv, nullifier, rk))
     }
 }
```
Comment (nuttycom) [inline zcash_primitives/src/transaction/components/sapling.rs:141]: It's weird to have this `impl OutputDescription<GrothProofBytes>` here when neither `OutputDescription` nor `GrothProofBytes` are defined in this module. I'm surprised it doesn't result in an orphan rule violation now, but it certainly will once the `sapling` module is moved out to the `sapling_crypto` crate, so I think another commit fixing this putative orphan rule violation would be good to include with this PR.

File: zcash_primitives/src/transaction/components/sapling.rs [hunk 2]
```diff
@@ -694,45 +163,34 @@ impl OutputDescription<GrothProofBytes> {
...
-    cv: ValueCommitment,
-    cmu: ExtractedNoteCommitment,
-    ephemeral_key: EphemeralKeyBytes,
-    enc_ciphertext: [u8; 580],
-    out_ciphertext: [u8; 80],
-}
-
-memuse::impl_no_dynamic_usage!(OutputDescriptionV5);
-
 impl OutputDescriptionV5 {
```
Comment (nuttycom) [inline zcash_primitives/src/transaction/components/sapling.rs:194]: Same comment with this impl.

**Review by nuttycom — CHANGES_REQUESTED** [review]
I think that the soon-to-be orphan-rule-violating `impl`s in `transaction::components::sapling` should be fixed as part of this PR.

File: zcash_primitives/src/transaction/components/sapling.rs (hunk 2 above)
Comment (str4d) [reply zcash_primitives/src/transaction/components/sapling.rs:194]: I was planning to change these in a separate PR, to keep the diff on this PR smaller.

**Review by nuttycom — APPROVED** [review]
utACK 043ad55

**Review by nuttycom — APPROVED** [review]
utACK c6dd9ad858f3efadde50f30edf9072ee7a6882f6

File: zcash_primitives/src/sapling/bundle.rs [hunk 3]
```diff
@@ -0,0 +1,736 @@
...
+    pub(crate) fn cv_mut(&mut self) -> &mut ValueCommitment {
+        &mut self.cv
+    }
+    pub(crate) fn cmu_mut(&mut self) -> &mut ExtractedNoteCommitment {
+        &mut self.cmu
+    }
+    pub(crate) fn ephemeral_key_mut(&mut self) -> &mut EphemeralKeyBytes {
+        &mut self.ephemeral_key
+    }
+    pub(crate) fn enc_ciphertext_mut(&mut self) -> &mut [u8; 580] {
```
Comment (daira) [inline zcash_primitives/src/sapling/bundle.rs:518]: Could this be `[u8; ENC_CIPHERTEXT_SIZE]`?

File: zcash_primitives/src/sapling/bundle.rs [hunk 4]
```diff
@@ -0,0 +1,736 @@
...
+        &mut self.enc_ciphertext
+    }
+    pub(crate) fn out_ciphertext_mut(&mut self) -> &mut [u8; 80] {
```
Comment (daira) [inline zcash_primitives/src/sapling/bundle.rs:521]: Could this be `[u8; OUT_CIPHERTEXT_SIZE]`?

Comment (daira) [reply zcash_primitives/src/sapling/bundle.rs:521]: (and similarly for other instances of these constants above)

**Pull request merged**